config/config.go

```diff
@@ -747,29 +785,81 @@ func (c *Config) Validate() error {
 						continue
 					}
 
-					errs = append(errs, fmt.Errorf(
-						"%s: value for 'description' must be string",
-						o.Name))
+					diags = diags.Append(fmt.Errorf(
+						"output %q: value for 'description' must be string",
+						o.Name,
+					))
 					continue
 				}
 				invalidKeys = append(invalidKeys, k)
 			}
 			if len(invalidKeys) > 0 {
-				errs = append(errs, fmt.Errorf(
-					"%s: output has invalid keys: %s",
-					o.Name, strings.Join(invalidKeys, ", ")))
+				diags = diags.Append(fmt.Errorf(
+					"output %q: invalid keys: %s",
+					o.Name, strings.Join(invalidKeys, ", "),
+				))
 			}
 			if !valueKeyFound {
-				errs = append(errs, fmt.Errorf(
-					"%s: output is missing required 'value' key", o.Name))
+				diags = diags.Append(fmt.Errorf(
+					"output %q: missing required 'value' argument", o.Name,
+				))
 			}
 
 			for _, v := range o.RawConfig.Variables {
 				if _, ok := v.(*CountVariable); ok {
-					errs = append(errs, fmt.Errorf(
-						"%s: count variables are only valid within resources", o.Name))
+					diags = diags.Append(fmt.Errorf(
+						"output %q: count variables are only valid within resources",
+						o.Name,
+					))
 				}
 			}
+
+			// Detect a common mistake of using a "count"ed resource in
+			// an output value without using the splat or index form.
+			// Prior to 0.11 this error was silently ignored, but outputs
+			// now have their errors checked like all other contexts.
+			//
+			// TODO: Remove this in 0.12.
+			for _, v := range o.RawConfig.Variables {
+				rv, ok := v.(*ResourceVariable)
+				if !ok {
+					continue
+				}
+
+				// If the variable seems to be treating the referenced
+				// resource as a singleton (no count specified) then
+				// we'll check to make sure it is indeed a singleton.
+				// It's a warning if not.
+
+				if rv.Multi || rv.Index != 0 {
+					// This reference is treating the resource as a
+					// multi-resource, so the warning doesn't apply.
+					continue
+				}
+
+				for _, r := range c.Resources {
+					if r.Id() != rv.ResourceId() {
+						continue
+					}
+
+					// We test specifically for the raw string "1" here
+					// because we _do_ want to generate this warning if
+					// the user has provided an expression that happens
+					// to return 1 right now, to catch situations where
+					// a count might dynamically be set to something
+					// other than 1 and thus splat syntax is still needed
+					// to be safe.
+					if r.RawCount != nil && r.RawCount.Raw != nil && r.RawCount.Raw["count"] != "1" {
```

Comment:
Since we're only going to conditionally returns errors and can be a little overly ambitious, do we want to catch this with a literal "1" too since that is also something the user could change? I think of there's a count value at all, it indicates that could be changes, otherwise it would have been left out.

Reply:
Oh I see, that's going to always have a raw value. I guess we should skip that rather than mess with the loader too.

Reply:
If it were straightforward to recognize presence vs. absense of `count` then I _would've_ done that here, but this compromise seems okay because the goal of this warning is to catch cases where an error might result from a change of _input variables_, rather than from a change to the module's own configuration.

As you saw, right now we can't distinguish these cases due to a default applied in the loader, and indeed I didn't want to fuss with that right now since I know from earlier HCL2 work that there is logic in core depending on that default.